packages/react-urql/src/hooks/useFragment.ts

```diff
@@ -0,0 +1,262 @@
+ * waiting for a result via {@link UseFragmentState.fetching}.
+ *
+ * @example
+ * ```ts
+ * import { gql, useFragment } from 'urql';
+ *
+ * const TodoFields = gql`
+ *   fragment TodoFields on Todo { id name }
+ * `;
+ *
+ * const Todo = (props) => {
+ *   const result = useQuery({
+ *     data: props.todo,
+ *     query: TodoFields,
+ *     variables: {},
+ *   });
+ *   // ...
+ * };
+ * ```
+ */
+export function useFragment<
+  Data extends Record<string, any> = Record<string, any>,
+>(args: UseFragmentArgs<Data>): UseFragmentState<Data> {
+  const client = useClient();
+  const cache = getCacheForClient(client);
+  const suspense = isSuspense(client, args.context);
+  const request = useRequest(args.query, args.data);
+
+  const getSnapshot = React.useCallback(
+    (
+      request: GraphQLRequest<Data, AnyVariables>,
+      data: Data,
+      suspense: boolean
+    ): Partial<UseFragmentState<Data>> => {
+      const cached = cache.get(request.key);
+      if (!cached) {
+        const fragment = request.query.definitions.find(
+          x =>
+            x.kind === Kind.FRAGMENT_DEFINITION &&
+            ((args.name && x.name.value === args.name) || !args.name)
+        ) as FragmentDefinitionNode | undefined;
+
+        if (!fragment) {
+          throw new Error(
+            'Passed document did not contain a fragment definition' + args.name
+              ? ` for ${args.name}`
+              : ''
+          );
+        }
+
+        const newResult = maskFragment<Data>(data, fragment.selectionSet);
+        if (newResult.fulfilled) {
+          cache.set(request.key, newResult.data as any);
+          return { data: newResult.data as any, fetching: false };
+        } else if (suspense) {
+          const promise = new Promise(() => {});
+          cache.set(request.key, promise);
+          throw promise;
+        } else {
+          return { fetching: true, data: newResult.data };
+        }
+      } else if (suspense && cached != null && 'then' in cached) {
+        throw cached;
+      }
+
+      return { fetching: false, data: (cached as OperationResult).data };
+    },
+    [cache, request]
+  );
+
+  const deps = [client, request, args.context, args.data] as const;
+
+  const [state, setState] = React.useState(
+    () =>
+      [
+        computeNextState(
+          initialState,
+          getSnapshot(request, args.data, suspense)
+        ),
+        deps,
+      ] as const
+  );
+
+  let currentResult = state[0];
+  if (hasDepsChanged(state[1], deps)) {
+    setState([
+      (currentResult = computeNextState(
+        state[0],
+        getSnapshot(request, args.data, suspense)
+      )),
+      deps,
+    ]);
+  }
+
+  return currentResult;
+}
+
+const maskFragment = <Data extends Record<string, any>>(
+  data: Data,
+  selectionSet: SelectionSetNode
+): { data: Data; fulfilled: boolean } => {
+  const maskedData = {};
+  let isDataComplete = true;
+  selectionSet.selections.forEach(selection => {
```

Comment:
There are a few things bothering me about doing that, we should def do `FragmentSpread` as alluded to in the todo, however what bothers me is that these spreads could be suspended themselves. I wanted to create the assumption that a `FragmentSpread` with `@defer` wouldn't be registered in `useFragment` but there would be an expectation that this is passed on to a new component.

For `InlineFragment` I have implemented it similar to `Field` and `defer` would trigger internal to the component.

In writing this I realised that we actually can't do `include`/`skip` as it stands because we don't have access to the variables the parent-operation was executed with, this could be a limitation for now but weakens the heuristic we use to see whether something is deferred significantly.